From 8f519e5e5801475e62e5b8364c6668f3acd78614 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Fri, 26 Aug 2022 13:58:13 +0300
Subject: [PATCH] [SW-86869] common: add range check for img desc pointer

Once manifest image is downloaded ROM performs sanity check of each
image downloaded as part of manifest. In order to get start address
of each image, rom adds base addr with the address from array received
as part manifest header. Currently ROM directly access that image addr
by adding the offset which might point to some other address outside
the manifest range. This patch performs range check for each image
descriptor base address to be within manifest downloaded area.

Change-Id: I44b9c60f1867e8d687eb754d3e447b5363958fb6
---

diff --git a/arch/gaudi3/arc/progress_bits.h b/arch/gaudi3/arc/progress_bits.h
index 487c10d..64cc6c2 100644
--- a/arch/gaudi3/arc/progress_bits.h
+++ b/arch/gaudi3/arc/progress_bits.h
@@ -16,133 +16,135 @@
 	PROGRESS_INIT_SPI_BOOT_DESC_DMA_ERROR = 9,
 	PROGRESS_INIT_SPI_BOOT_DESC_DMA_FINISH = 10,
 	PROGRESS_INIT_SPI_BOOT_DESC_HDR_MAGIC_FAIL = 11,
-	PROGRESS_INIT_SPI_BOOT_DESC_VERIFICATION_FAIL = 12,
-	PROGRESS_INIT_SPI_BOOT_DESC_INFO_SET_FAIL = 13,
-	PROGRESS_INIT_SPI_BOOT_DESC_DATA_SET_FAIL = 14,
-	PROGRESS_INIT_SPI_BOOT_DESC_INFO_GET_FAIL = 15,
-	PROGRESS_INIT_SPI_BOOT_DESC_DATA_GET_FAIL = 16,
-	PROGRESS_INIT_SPI_BOOT_DESC_FAIL = 17,
-	PROGRESS_INIT_EFUSE_SPI = 18,
-	PROGRESS_INIT_EFUSE_SPI_SLV_TIMEDOUT = 19,
-	PROGRESS_REVOKE_LOC = 20,
-	PROGRESS_REVOKE_LOC_0 = 21,
-	PROGRESS_REVOKE_LOC_1 = 22,
-	PROGRESS_KEY_LOC = 23,
-	PROGRESS_KEY_LOC_0 = 24,
-	PROGRESS_KEY_LOC_1 = 25,
-	PROGRESS_VER_LOC = 26,
-	PROGRESS_VER_LOC_0 = 27,
-	PROGRESS_VER_LOC_1 = 28,
-	PROGRESS_SPI_BOOT_DESC_VERIFY_START = 29,
-	PROGRESS_SPI_BOOT_DESC_VERIFY_VERSION_FAIL = 30,
-	PROGRESS_SPI_BOOT_DESC_VERIFY_KEY_REVOKED = 31,
-	PROGRESS_SPI_BOOT_DESC_VERIFY_KEY_HASH_FAIL = 32,
-	PROGRESS_SPI_BOOT_DESC_VERIFY_AUTH_FAIL = 33,
-	PROGRESS_SPI_BOOT_DESC_VERIFY_SUCCESS = 34,
-	PROGRESS_BRINGUP_PCI_PRI = 35,
-	PROGRESS_BRINGUP_PCI_PRI_TOO_BIG = 36,
-	PROGRESS_BRINGUP_PCI_PRI_CRC_FAIL = 37,
-	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_DEST_ADDR = 38,
-	PROGRESS_BRINGUP_PCI_PRI_INVALID_IMG_SIZE = 39,
-	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_IMG_SIZE = 40,
-	PROGRESS_BRINGUP_PCI_PRI_DMA_ERROR = 41,
-	PROGRESS_BRINGUP_PCI_PRI_DMA = 42,
-	PROGRESS_BRINGUP_PCI_PRI_KEY_REVOKED = 43,
-	PROGRESS_BRINGUP_PCI_PRI_KEY_AUTH_FAIL = 44,
-	PROGRESS_BRINGUP_PCI_PRI_SECURITY = 45,
-	PROGRESS_BRINGUP_PCI_SEC = 46,
-	PROGRESS_BRINGUP_PCI_SEC_TOO_BIG = 47,
-	PROGRESS_BRINGUP_PCI_SEC_CRC_FAIL = 48,
-	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_DEST_ADDR = 49,
-	PROGRESS_BRINGUP_PCI_SEC_INVALID_IMG_SIZE = 50,
-	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_IMG_SIZE = 51,
-	PROGRESS_BRINGUP_PCI_SEC_DMA_ERROR = 52,
-	PROGRESS_BRINGUP_PCI_SEC_DMA = 53,
-	PROGRESS_BRINGUP_PCI_SEC_KEY_REVOKED = 54,
-	PROGRESS_BRINGUP_PCI_SEC_KEY_AUTH_FAIL = 55,
-	PROGRESS_BRINGUP_PCI_SEC_SECURITY = 56,
-	PROGRESS_ARCBOOT_IN_RECOVERY = 57,
-	PROGRESS_ARCBOOT_ROM_TPM = 58,
-	PROGRESS_ARCBOOT_ROM_TPM_INIT_FAIL = 59,
-	PROGRESS_ARCBOOT_ROM_TPM_STARTUP_FAIL = 60,
-	PROGRESS_ARCBOOT_ROM_TPM_PCR_EXTEND_FAIL = 61,
-	PROGRESS_ARCBOOT_ROM_TPM_FINISH = 62,
-	PROGRESS_BRINGUP_TPM_SPI = 63,
-	PROGRESS_BRINGUP_TPM_SPI_PRI = 64,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_TOO_BIG = 65,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_CRC_FAIL = 66,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_UNALIGNED_DEST_ADDR = 67,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_INVALID_IMG_SIZE = 68,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_UNALIGNED_IMG_SIZE = 69,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_DMA_ERROR = 70,
+	PROGRESS_INIT_SPI_BOOT_DESC_INVALID_IMG_OFFSET = 12,
+	PROGRESS_INIT_SPI_BOOT_DESC_VERIFICATION_FAIL = 13,
+	PROGRESS_INIT_SPI_BOOT_DESC_INFO_SET_FAIL = 14,
+	PROGRESS_INIT_SPI_BOOT_DESC_DATA_SET_FAIL = 15,
+	PROGRESS_INIT_SPI_BOOT_DESC_INFO_GET_FAIL = 16,
+	PROGRESS_INIT_SPI_BOOT_DESC_DATA_GET_FAIL = 17,
+	PROGRESS_INIT_SPI_BOOT_DESC_FAIL = 18,
+	PROGRESS_INIT_EFUSE_SPI = 19,
+	PROGRESS_INIT_EFUSE_SPI_SLV_TIMEDOUT = 20,
+	PROGRESS_REVOKE_LOC = 21,
+	PROGRESS_REVOKE_LOC_0 = 22,
+	PROGRESS_REVOKE_LOC_1 = 23,
+	PROGRESS_KEY_LOC = 24,
+	PROGRESS_KEY_LOC_0 = 25,
+	PROGRESS_KEY_LOC_1 = 26,
+	PROGRESS_VER_LOC = 27,
+	PROGRESS_VER_LOC_0 = 28,
+	PROGRESS_VER_LOC_1 = 29,
+	PROGRESS_SPI_BOOT_DESC_VERIFY_START = 30,
+	PROGRESS_SPI_BOOT_DESC_VERIFY_VERSION_FAIL = 31,
+	PROGRESS_SPI_BOOT_DESC_VERIFY_KEY_REVOKED = 32,
+	PROGRESS_SPI_BOOT_DESC_VERIFY_KEY_HASH_FAIL = 33,
+	PROGRESS_SPI_BOOT_DESC_VERIFY_AUTH_FAIL = 34,
+	PROGRESS_SPI_BOOT_DESC_VERIFY_SUCCESS = 35,
+	PROGRESS_BRINGUP_PCI_PRI = 36,
+	PROGRESS_BRINGUP_PCI_PRI_TOO_BIG = 37,
+	PROGRESS_BRINGUP_PCI_PRI_CRC_FAIL = 38,
+	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_DEST_ADDR = 39,
+	PROGRESS_BRINGUP_PCI_PRI_INVALID_IMG_SIZE = 40,
+	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_IMG_SIZE = 41,
+	PROGRESS_BRINGUP_PCI_PRI_DMA_ERROR = 42,
+	PROGRESS_BRINGUP_PCI_PRI_DMA = 43,
+	PROGRESS_BRINGUP_PCI_PRI_KEY_REVOKED = 44,
+	PROGRESS_BRINGUP_PCI_PRI_KEY_AUTH_FAIL = 45,
+	PROGRESS_BRINGUP_PCI_PRI_SECURITY = 46,
+	PROGRESS_BRINGUP_PCI_SEC = 47,
+	PROGRESS_BRINGUP_PCI_SEC_TOO_BIG = 48,
+	PROGRESS_BRINGUP_PCI_SEC_CRC_FAIL = 49,
+	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_DEST_ADDR = 50,
+	PROGRESS_BRINGUP_PCI_SEC_INVALID_IMG_SIZE = 51,
+	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_IMG_SIZE = 52,
+	PROGRESS_BRINGUP_PCI_SEC_DMA_ERROR = 53,
+	PROGRESS_BRINGUP_PCI_SEC_DMA = 54,
+	PROGRESS_BRINGUP_PCI_SEC_KEY_REVOKED = 55,
+	PROGRESS_BRINGUP_PCI_SEC_KEY_AUTH_FAIL = 56,
+	PROGRESS_BRINGUP_PCI_SEC_SECURITY = 57,
+	PROGRESS_ARCBOOT_IN_RECOVERY = 58,
+	PROGRESS_ARCBOOT_ROM_TPM = 59,
+	PROGRESS_ARCBOOT_ROM_TPM_INIT_FAIL = 60,
+	PROGRESS_ARCBOOT_ROM_TPM_STARTUP_FAIL = 61,
+	PROGRESS_ARCBOOT_ROM_TPM_PCR_EXTEND_FAIL = 62,
+	PROGRESS_ARCBOOT_ROM_TPM_FINISH = 63,
+	PROGRESS_BRINGUP_TPM_SPI = 64,
+	PROGRESS_BRINGUP_TPM_SPI_PRI = 65,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_TOO_BIG = 66,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_CRC_FAIL = 67,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_UNALIGNED_DEST_ADDR = 68,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_INVALID_IMG_SIZE = 69,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_UNALIGNED_IMG_SIZE = 70,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_DMA_ERROR = 71,
 	PROGRESS_BRINGUP_TPM_SPI_PRI_DMA = 72,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_KEY_REVOKED = 72,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_KEY_AUTH_FAIL = 73,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_SECURITY = 74,
-	PROGRESS_BRINGUP_TPM_SPI_SEC = 75,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_TOO_BIG = 76,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_CRC_FAIL = 77,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_UNALIGNED_DEST_ADDR = 78,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_INVALID_IMG_SIZE = 79,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_UNALIGNED_IMG_SIZE = 80,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_DMA_ERROR = 81,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_DMA = 82,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_KEY_REVOKED = 83,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_KEY_AUTH_FAIL = 84,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_SECURITY = 85,
-	PROGRESS_ARCBOOT_TPM_SPI_FINISH = 86,
-	PROGRESS_ARCBOOT_TPM_SPI_FAIL = 87,
-	PROGRESS_PPBOOT_START = 88,
-	PROGRESS_PPBOOT_FINISH = 89,
-	PROGRESS_BRINGUP_RAM_PRI = 90,
-	PROGRESS_BRINGUP_RAM_PRI_DMA = 91,
-	PROGRESS_BRINGUP_RAM_PRI_SECURITY = 92,
-	PROGRESS_BRINGUP_RAM_SEC = 93,
-	PROGRESS_BRINGUP_RAM_SEC_DMA = 94,
-	PROGRESS_BRINGUP_RAM_SEC_SECURITY = 95,
-	PROGRESS_ARCBOOT_RAM_START = 96,
-	PROGRESS_ARCBOOT_RECOVERY = 97,
-	PROGRESS_ARCBOOT_RECOVERY_DMA = 98,
-	PROGRESS_ARCBOOT_IN_ROM_RECOVERY = 99,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_START = 100,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_INIT = 101,
-	PROGRESS_ARCBOOT_ROM_PCIE_PHY_SRAM_INIT_TIMEDOUT = 102,
-	PROGRESS_ARCBOOT_ROM_PCIE_PHY_PARA_STS_CLR_TIMEDOUT = 103,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_LIMIT_GEN1 = 104,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_IATU_CONFIG = 105,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_FINISH = 106,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_START = 107,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_STS_BUSY = 108,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_TIMEDOUT = 109,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_SUCCESS = 110,
-	PROGRESS_ARCBOOT_ROM_RECOVERY_END = 111,
-	PROGRESS_ARCBOOT_IN_EROM_RECOVERY = 112,
-	PROGRESS_INIT_EROM_BOOT_DESC_EXCEEDS_MAX_SIZE = 113,
-	PROGRESS_INIT_EROM_BOOT_DESC_DMA_START = 114,
-	PROGRESS_INIT_EROM_BOOT_DESC_CRC_FAIL = 115,
-	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_DEST = 116,
-	PROGRESS_INIT_EROM_BOOT_DESC_INVALID_IMG_SIZE = 117,
-	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_IMG_SIZE = 118,
-	PROGRESS_INIT_EROM_BOOT_DESC_DMA_ERROR = 119,
-	PROGRESS_INIT_EROM_BOOT_DESC_DMA_FINISH = 120,
-	PROGRESS_INIT_EROM_BOOT_DESC_HDR_MAGIC_FAIL = 121,
-	PROGRESS_INIT_EROM_BOOT_DESC_VERIFICATION_FAIL = 122,
-	PROGRESS_INIT_EROM_BOOT_DESC_FAIL = 123,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_START = 124,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_VERSION_FAIL = 125,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_REVOKED = 126,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_HASH_FAIL = 127,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_AUTH_FAIL = 128,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_SUCCESS = 129,
-	PROGRESS_ARCBOOT_EROM_IMG0 = 130,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_START = 131,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_CRC_FAILED = 132,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_DEST = 133,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_INVALID_IMG_SIZE = 134,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_IMG_SIZE = 135,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_ERROR = 136,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_FINISH = 137,
-	PROGRESS_ARCBOOT_EROM_IMG0_VERIFICATION_FAIL = 138,
-	PROGRESS_ARCBOOT_EROM_IMG0_START = 139,
-	PROGRESS_ARCBOOT_EROM_IMG0_FINISH = 140,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_KEY_REVOKED = 73,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_KEY_AUTH_FAIL = 74,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_SECURITY = 75,
+	PROGRESS_BRINGUP_TPM_SPI_SEC = 76,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_TOO_BIG = 77,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_CRC_FAIL = 78,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_UNALIGNED_DEST_ADDR = 79,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_INVALID_IMG_SIZE = 80,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_UNALIGNED_IMG_SIZE = 81,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_DMA_ERROR = 82,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_DMA = 83,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_KEY_REVOKED = 84,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_KEY_AUTH_FAIL = 85,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_SECURITY = 86,
+	PROGRESS_ARCBOOT_TPM_SPI_FINISH = 87,
+	PROGRESS_ARCBOOT_TPM_SPI_FAIL = 88,
+	PROGRESS_PPBOOT_START = 89,
+	PROGRESS_PPBOOT_FINISH = 90,
+	PROGRESS_BRINGUP_RAM_PRI = 91,
+	PROGRESS_BRINGUP_RAM_PRI_DMA = 92,
+	PROGRESS_BRINGUP_RAM_PRI_SECURITY = 93,
+	PROGRESS_BRINGUP_RAM_SEC = 94,
+	PROGRESS_BRINGUP_RAM_SEC_DMA = 95,
+	PROGRESS_BRINGUP_RAM_SEC_SECURITY = 96,
+	PROGRESS_ARCBOOT_RAM_START = 97,
+	PROGRESS_ARCBOOT_RECOVERY = 98,
+	PROGRESS_ARCBOOT_RECOVERY_DMA = 99,
+	PROGRESS_ARCBOOT_IN_ROM_RECOVERY = 100,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_START = 101,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_INIT = 102,
+	PROGRESS_ARCBOOT_ROM_PCIE_PHY_SRAM_INIT_TIMEDOUT = 103,
+	PROGRESS_ARCBOOT_ROM_PCIE_PHY_PARA_STS_CLR_TIMEDOUT = 104,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_LIMIT_GEN1 = 105,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_IATU_CONFIG = 106,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_FINISH = 107,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_START = 108,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_STS_BUSY = 109,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_TIMEDOUT = 110,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_SUCCESS = 111,
+	PROGRESS_ARCBOOT_ROM_RECOVERY_END = 112,
+	PROGRESS_ARCBOOT_IN_EROM_RECOVERY = 113,
+	PROGRESS_INIT_EROM_BOOT_DESC_EXCEEDS_MAX_SIZE = 114,
+	PROGRESS_INIT_EROM_BOOT_DESC_DMA_START = 115,
+	PROGRESS_INIT_EROM_BOOT_DESC_CRC_FAIL = 116,
+	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_DEST = 117,
+	PROGRESS_INIT_EROM_BOOT_DESC_INVALID_IMG_SIZE = 118,
+	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_IMG_SIZE = 119,
+	PROGRESS_INIT_EROM_BOOT_DESC_DMA_ERROR = 120,
+	PROGRESS_INIT_EROM_BOOT_DESC_DMA_FINISH = 121,
+	PROGRESS_INIT_EROM_BOOT_DESC_HDR_MAGIC_FAIL = 122,
+	PROGRESS_INIT_EROM_BOOT_DESC_INVALID_IMG_OFFSET = 123,
+	PROGRESS_INIT_EROM_BOOT_DESC_VERIFICATION_FAIL = 124,
+	PROGRESS_INIT_EROM_BOOT_DESC_FAIL = 125,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_START = 126,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_VERSION_FAIL = 127,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_REVOKED = 128,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_HASH_FAIL = 129,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_AUTH_FAIL = 130,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_SUCCESS = 131,
+	PROGRESS_ARCBOOT_EROM_IMG0 = 132,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_START = 133,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_CRC_FAILED = 134,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_DEST = 135,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_INVALID_IMG_SIZE = 136,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_IMG_SIZE = 137,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_ERROR = 138,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_FINISH = 139,
+	PROGRESS_ARCBOOT_EROM_IMG0_VERIFICATION_FAIL = 140,
+	PROGRESS_ARCBOOT_EROM_IMG0_START = 141,
+	PROGRESS_ARCBOOT_EROM_IMG0_FINISH = 142,
 };
diff --git a/common/boot_desc.c b/common/boot_desc.c
index 91b9f97..e2b21f5 100644
--- a/common/boot_desc.c
+++ b/common/boot_desc.c
@@ -127,6 +127,14 @@
 	for (i = 0; i < p_boot_ptr->count; i++) {
 		p_img_desc = (struct hl_fw_img_desc *)(p_img_ptr_ctx->dest_addr_arc + p_offset[i]);
 
+		/* Range check for p_img_desc */
+		if (!(((uint32_t) p_img_desc >= p_img_ptr_ctx->dest_addr_arc) &&
+			((uint32_t) p_img_desc < p_img_ptr_ctx->dest_addr_arc + img_ptrs_size))) {
+			arcboot_rom_status_update_bringup(dma_sts_base,
+				PROGRESS_BOOT_DESC_INVALID_IMG_OFFSET);
+			return 0;
+		}
+
 		if (!boot_desc_img_desc_is_valid(p_img_desc)) {
 			arcboot_rom_status_update_bringup(dma_sts_base, PROGRESS_BOOT_DESC_VERIFICATION_FAIL);
 			return 0;
diff --git a/include/arcboot.h b/include/arcboot.h
index a14ccad..880baee 100644
--- a/include/arcboot.h
+++ b/include/arcboot.h
@@ -117,6 +117,7 @@
 	PROGRESS_DMA_ERROR,
 	PROGRESS_DMA_FINISH,
 	PROGRESS_BOOT_DESC_HDR_MAGIC_FAIL,
+	PROGRESS_BOOT_DESC_INVALID_IMG_OFFSET,
 	PROGRESS_BOOT_DESC_VERIFICATION_FAIL,
 };
 
